CodeQL: Python topics (2166) - WIP

felicitymay opened the pull request:
This PR contains changes that start the work of preparing the CodeQL for Python topics for future migration, as outlined in product-documentation issue 2166.

## Notes

For most topics, this PR makes just the following changes:

* Create an intro.
* Update some headings for better internal consistency and consistency with content models. 
* Update some links for localization and content model guidelines (ignoring all links to items in the QL library reference which should not be translated).
* Change "What's next" section into a "Further reading" section.

Original | New | Notes
---------|-------|-------
[Tutorial: Control flow analysis](https://help.semmle.com/QL/learn-ql/python/control-flow.html)|[Analyzing control flow in Python](http://docteam.internal.semmle.com/felicity/2166-python/python/control-flow.html)| In addition to the above, merged an image of a control flow graph into this file (previously in a separate topic).
[Tutorial: Functions](https://help.semmle.com/QL/learn-ql/python/functions.html)|[Functions in Python](http://docteam.internal.semmle.com/felicity/2166-python/python/functions.html)| ---
[Introducing the CodeQL libraries for Python](https://help.semmle.com/QL/learn-ql/python/introduce-libraries-python.html)|[CodeQL library for Python](http://docteam.internal.semmle.com/felicity/2166-python/python/introduce-libraries-python.html)| ---
[Tutorial: Points-to analysis and type inference](https://help.semmle.com/QL/learn-ql/python/pointsto-type-infer.html)|[Pointer analysis and type inference in Python](http://docteam.internal.semmle.com/felicity/2166-python/python/pointsto-type-infer.html)|---
[CodeQL for Python](https://help.semmle.com/QL/learn-ql/python/ql-for-python.html)|[CodeQL for Python](http://docteam.internal.semmle.com/felicity/2166-python/python/ql-for-python.html)| This map topic now contains only its `intro` and a toc of the child articles. The other resources are moved into a [reusable snippet](https://github.com/Semmle/ql/pull/2866) for use in the "Further reading" section of all child topics.
[Tutorial: Statements and expressions](https://help.semmle.com/QL/learn-ql/python/statements-expressions.html)|[Expressions and statements in Python](http://docteam.internal.semmle.com/felicity/2166-python/python/statements-expressions.html)|---
[Taint tracking and data flow analysis in Python](https://help.semmle.com/QL/learn-ql/python/taint-tracking.html)|[Analyzing data flow and tracking tainted data in Python](http://docteam.internal.semmle.com/felicity/2166-python/python/taint-tracking.html)|---

## Questions

1. There is a gap between the links in the "Further reading" section that are directly defined and those that are in the resuable snippet. This seems to be unavoidable with Sphinx unless we reverse the order, but the links defined in the topic are more important so I think they should be listed first. I think the gap is acceptable in the short term and better than the alternatives. Happy to discuss if you disagree.
1. What do you think of this style of `intro` - review comments very welcome.

felicitymay commented:
A possible flaw with this approach to the introductions - the sidebar has both the new titles and the old titles for the child articles. This may simply be because I haven't updated the titles in the articles yet. I'll look into it on Friday.

shati-patel commented:
> A possible flaw with this approach to the introductions - the sidebar has both the new titles and the old titles for the child articles. This may simply be because I haven't updated the titles in the articles yet. I'll look into it on Friday.

This is probably because the sidebar entries are autogenerated from the elements in `toctree` (currently using old titles) **and** the top-level headers in the file (using new titles). Even if you update the titles in the articles, they'll still appear twice! 
(I know this Sphinx formatting won't be preserved when we migrate the docs, but if you want things to look nice here you could put the articles back in a list instead of as headers?)

felicitymay commented:
Thanks for the tip Shati - much appreciated.

I've tried the page as a bulleted list and it looks rather cramped to me: http://docteam.internal.semmle.com/felicity/2166-python-bullets/python/ql-for-python.html.
What do you think of: http://docteam.internal.semmle.com/felicity/2166-python/python/ql-for-python.html (which uses a [`rubric` directive](http://www.sphinx-doc.org/en/stable/usage/restructuredtext/directives.html#directive-rubric))?

shati-patel commented:
> I've tried the page as a bulleted list and it looks rather cramped to me: http://docteam.internal.semmle.com/felicity/2166-python-bullets/python/ql-for-python.html.
> What do you think of: http://docteam.internal.semmle.com/felicity/2166-python/python/ql-for-python.html (which uses a [`rubric` directive](http://www.sphinx-doc.org/en/stable/usage/restructuredtext/directives.html#directive-rubric))?

Agreed, the bulleted list looks pretty cramped, but the second version looks great! 🎉 (Though  all of this stuff will be autogenerated once we use the GitHub process, so I guess it doesn't matter _too_ much?)

felicitymay commented:
> (Though all of this stuff will be autogenerated once we use the GitHub process, so I guess it doesn't matter too much?)

That's a good point. I think this might be worth talking about. Since the migration isn't expected to start soon, it might be nice to make the pre-migration changes in such a way that we can merge the `docs-preparation` branch back into `master` once we've finished the changes. I suspect that we might all have different views on what the resulting output should look like.

felicitymay commented:
Following discussions with @shati-patel and @jf205 - I removed the duplicated introductions from the map topic and uncommented the `toctree`.

felicitymay commented:
As discussed, I've backed out the wider-reaching changes that I'd started in the Library introduction topic. I've updated the description to match the commits that are now in this PR and the preview is also updated.

I'd be very grateful for a review.

shati-patel commented on docs/language/learn-ql/python/taint-tracking.rst:
```diff
@@ -1,8 +1,10 @@
 
-Overview
---------
+You can use CodeQL to track the flow of data through a Python program to its use. Tracking user-controlled, or tainted, data is a key technique for security researchers.
```
```suggestion
You can use CodeQL to track the flow of data through a Python program. Tracking user-controlled, or tainted, data is a key technique for security researchers.
```

felicitymay commented on docs/language/learn-ql/python/ql-for-python.rst:
```diff
@@ -1,37 +1,16 @@
 .. toctree::
    :glob:
-   :hidden:
+   :maxdepth: 2
```
I agree about the `maxdepth` but thought that, in the absence of introductions on this page, the top-level headings at least gave more information about what was in each topic.

Happy to change to `1` if that's preferred though.

felicitymay commented on docs/language/learn-ql/python/control-flow.rst:
```diff
@@ -35,7 +44,8 @@ The simplest use of the ``ControlFlowNode`` and ``AstNode`` classes is to find u
 ➤ `See this in the query console <https://lgtm.com/query/669220024/>`__. The demo projects on LGTM.com all have some code that has no control flow node, and is therefore unreachable. However, since the ``Module`` class is also a subclass of the ``AstNode`` class, the query also finds any modules implemented in C or with no source code. Therefore, it is better to find all unreachable statements:
 
-**Unreachable statements**
+Example finding unreachable statements
```
Thanks - good point

felicitymay commented on docs/language/learn-ql/python/control-flow.rst:
```diff
@@ -35,7 +44,8 @@ The simplest use of the ``ControlFlowNode`` and ``AstNode`` classes is to find u
 
 ➤ `See this in the query console <https://lgtm.com/query/669220024/>`__. The demo projects on LGTM.com all have some code that has no control flow node, and is therefore unreachable. However, since the ``Module`` class is also a subclass of the ``AstNode`` class, the query also finds any modules implemented in C or with no source code. Therefore, it is better to find all unreachable statements:
```
```suggestion
➤ `See this in the query console <https://lgtm.com/query/669220024/>`__. The demo projects on LGTM.com all have some code that has no control flow node, and is therefore unreachable. However, since the ``Module`` class is also a subclass of the ``AstNode`` class, the query also finds any modules implemented in C or with no source code. Therefore, it is better to find all unreachable statements.
```

felicitymay commented on docs/language/learn-ql/python/functions.rst:
```diff
@@ -1,7 +1,9 @@
 ===================
 
-This example uses the standard CodeQL class ``Function`` (see :doc:`Introducing the Python libraries <introduce-libraries-python>`).
+Functions are key building blocks of Python code bases. You can find functions and identify calls to them using syntactic classes from the standard CodeQL library.
```
How about:
```suggestion
Functions are key building blocks of Python code bases. You can use syntactic classes from the standard CodeQL library to find functions and identify calls to them.
```

shati-patel replied:
^ I like that suggestion :-)

felicitymay commented on docs/language/learn-ql/python/functions.rst:
```diff
@@ -1,7 +1,9 @@
 ===================
 
-This example uses the standard CodeQL class ``Function`` (see :doc:`Introducing the Python libraries <introduce-libraries-python>`).
+Functions are key building blocks of Python code bases. You can find functions and identify calls to them using syntactic classes from the standard CodeQL library.
```
Hmm. I was trying to differentiate between the style of introductions for referential versus quickstart topics. It felt as if a referential topic ought to have more of a statement and less of a call to action. 

However, looking at help.github.com, I can see that two of the three reference topics linked from the content models use "You can" to begin their introductions. https://help.github.com/en/actions/reference/workflow-syntax-for-github-actions is the only exception. So probably this isn't necessary. I can see that you and Alistair have both used the "you can" approach.

felicitymay replied:
I've created a Google doc to summarize the questions arising from the pre-migration work we've done so far. I've included some examples and will add this too.

felicitymay commented on docs/language/learn-ql/python/introduce-libraries-python.rst:
```diff
@@ -1,16 +1,16 @@
+=========================
 
-There is an extensive library for analyzing CodeQL databases extracted from Python projects. The classes in this library present the data from a database in an object-oriented form and provide abstractions and predicates to help you with common analysis tasks.  The library is implemented as a set of QL modules, that is, files with the extension ``.qll``. The module ``python.qll`` imports all the  ...
+Overview of the extensive library you use to analyze databases generated from Python code bases. This library uses classes with abstractions and predicates to present the data in an object-oriented form. This abstraction makes it easier for you to write queries.
```
```suggestion
This is an overview of the extensive library you use to analyze databases generated from Python code bases. Using this library makes it easier for you to write queries.
```

felicitymay commented on docs/language/learn-ql/python/introduce-libraries-python.rst:
```diff
@@ -1,16 +1,16 @@
+-----------------------------------
 
-   import python
+The CodeQL library for each programming language is implemented as a set of QL modules, that is, files with the extension ``.qll``. The module ``python.qll`` imports all the core Python library modules, so you can include the complete library by beginning your query with:
```
```suggestion
The CodeQL library for each programming language uses classes with abstractions and predicates to present data in an object-oriented form. This abstraction makes it easier for you to write queries.

Each CodeQL library is implemented as a set of QL modules, that is, files with the extension ``.qll``. The module ``python.qll`` imports all the core Python library modules, so you can include the complete library by beginning your query with:
```

felicitymay commented on docs/language/learn-ql/python/introduce-libraries-python.rst:
```diff
@@ -321,16 +327,21 @@ The CodeQL library for Python also supplies classes to specify taint-tracking an
 - `Configuration <https://help.semmle.com/qldoc/python/semmle/python/dataflow/Configuration.qll/type.Configuration$TaintTracking$Configuration.html>`__
 
-These classes are explained in more detail in :doc:`Tutorial: Taint tracking and data flow analysis in Python <taint-tracking>`.
+For more information about these classes, see ":doc:`Analyzing data flow and tracking tainted data in Python <taint-tracking>`".
```
Blast - I thought that I'd found all of those...

felicitymay commented on docs/language/learn-ql/python/control-flow.rst:
```diff
@@ -100,8 +111,9 @@ Combining these conditions we get:
 
 ➤ `See this in the query console <https://lgtm.com/query/671000028/>`__. This typically gives a very large number of results, because it is a common occurrence in normal control flow. It is, however, an example of the sort of control-flow analysis that is possible. Control-flow analyses such as this are an important aid to data flow analysis which is covered in the next tutorial.
```
```suggestion
➤ `See this in the query console <https://lgtm.com/query/671000028/>`__. This typically gives a very large number of results, because it is a common occurrence in normal control flow. It is, however, an example of the sort of control-flow analysis that is possible. Control-flow analyses such as this are an important aid to data flow analysis. For more information, see :doc:`Analyzing data flow and tracking tainted data in Python <taint-tracking>`.
```

shati-patel commented on docs/language/learn-ql/python/control-flow.rst:
```diff
@@ -1,7 +1,9 @@
+================================
 
-To analyze the `Control-flow graph <http://en.wikipedia.org/wiki/Control_flow_graph>`__ of a ``Scope`` we can use the two CodeQL classes ``ControlFlowNode`` and ``BasicBlock``. These classes allow you to ask such questions as "can you reach point A from point B?" or "Is it possible to reach point B *without* going through point A?". To report results we use the class ``AstNode``, which represents ...
+You can write CodeQL queries to explore the control flow graph of a Python program, for example, to discover unreachable code or mutually exclusive blocks of code.
```
```suggestion
You can write CodeQL queries to explore the control-flow graph of a Python program, for example, to discover unreachable code or mutually exclusive blocks of code.
```

shati-patel commented on docs/language/learn-ql/python/introduce-libraries-python.rst:
```diff
@@ -1,16 +1,18 @@
+-----------------------------------
 
-   import python
+The CodeQL library for each programming language uses classes with abstractions and predicates to present data in an object-oriented form. This abstraction makes it easier for you to write queries.
```
Perhaps remove the final sentence? That avoids duplicating "makes it easier for you to write queries."

felicitymay replied:
```suggestion
The CodeQL library for each programming language uses classes with abstractions and predicates to present data in an object-oriented form.
```

felicitymay commented:
I've updated the introductions following Emily's feedback in https://docs.google.com/document/d/1x917mlEHcpUBQqyKrQ7aogw8dG3FseO32m1s62fuIJA/edit.

I think that this is ready for a final review and merge.

felicitymay commented on docs/language/learn-ql/python/introduce-libraries-python.rst:
```diff
@@ -154,7 +154,7 @@ Both forms are equivalent. Using the positive expression, the whole query looks
 
 Summary
-^^^^^^^
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```
```suggestion
^^^^^^^
```

felicitymay commented:
@shati-patel - do you have any more feedback before this is merged?

jf205 commented:
@felicitymay: As with #2866, I think this is good to merge after you have looked at the most recent suggestions, pending approval from @shati-patel. Further changes can be made in a follow-up PR 👍

felicitymay commented on docs/language/learn-ql/python/introduce-libraries-python.rst:
```diff
@@ -1,16 +1,18 @@
+=========================
 
-There is an extensive library for analyzing CodeQL databases extracted from Python projects. The classes in this library present the data from a database in an object-oriented form and provide abstractions and predicates to help you with common analysis tasks.  The library is implemented as a set of QL modules, that is, files with the extension ``.qll``. The module ``python.qll`` imports all the  ...
+When you need to analyze a Python program, you can make use of the large collection of classes in the Python library for CodeQL.
```
Thanks for picking this up, but your suggestion has a typo. I suggest this:

```suggestion
When you need to analyze a Python program, you can make use of the large collection of classes in the CodeQL libraries for Python.
```

felicitymay commented on docs/language/learn-ql/python/introduce-libraries-python.rst:
```diff
@@ -1,16 +1,18 @@
-Introducing the CodeQL libraries for Python
-===========================================
+CodeQL library for Python
+=========================
```
Hmm. This suggests lots of other changes to this topic for consistency.

jf205 replied:
Stick with libraries. We'll update the other language topics accordingly.

felicitymay replied:
On balance, the more that I read this topic, the more convinced I am that it ought to be "library" - I think you're right that we should standardize on this, one way or the other. I'm going to leave this as-is for now pending that discussion.

jf205 replied:
On second thoughts, I think 'libraries' is consistent with what happens when you include `import python` in your query--although there is a single import statement, several `.qll` files (library files) are made available to use in the query. Not a big deal, but I agree we should  merge now, standardize later.

felicitymay replied:
Sorry - spotted this after I merged it. It's more than just a title though - we are inconsistent within the documentation in how we refer to things. I think of a language library being made up of a series of library files, each of which defines a library module.

felicitymay replied:
It seems as if I'm probably in a minority.

felicitymay commented:
Thanks for all the review comments. I'm happy to make any further changes in a follow up PR.
